src/firmware/main/math/polynomial_2d.c

```diff
@@ -0,0 +1,178 @@
+    float calculateL2NormAtValueOrder##N(Polynomial2dOrder##N##_t p, float t)            \
+    {                                                                                    \
+        const float x_value = shared_polynomial1d_getValueOrder##N(p.x, t);              \
+        const float y_value = shared_polynomial1d_getValueOrder##N(p.y, t);              \
+                                                                                         \
+        return sqrt(pow(x_value, 2.0) + pow(y_value, 2.0));                              \
+    }
+GENERATE_2D_POLYNOMIAL_CALCULATE_L2_NORM_DEFINITION(3);
+GENERATE_2D_POLYNOMIAL_CALCULATE_L2_NORM_DEFINITION(2);
+GENERATE_2D_POLYNOMIAL_CALCULATE_L2_NORM_DEFINITION(1);
+GENERATE_2D_POLYNOMIAL_CALCULATE_L2_NORM_DEFINITION(0);
+
+#define GENERATE_2D_POLYNOMIAL_GET_VALUE_FUNCTION_DEFINITION(N)                          \
+    Vector2d_t shared_polynomial2d_getValueOrder##N(Polynomial2dOrder##N##_t p, float t) \
+    {                                                                                    \
+        Vector2d_t result = {                                                            \
+            .x = shared_polynomial1d_getValueOrder##N(p.x, t),                           \
+            .y = shared_polynomial1d_getValueOrder##N(p.y, t),                           \
+        };                                                                               \
+        return result;                                                                   \
+    }
+GENERATE_2D_POLYNOMIAL_GET_VALUE_FUNCTION_DEFINITION(3)
+GENERATE_2D_POLYNOMIAL_GET_VALUE_FUNCTION_DEFINITION(2)
+GENERATE_2D_POLYNOMIAL_GET_VALUE_FUNCTION_DEFINITION(1)
+GENERATE_2D_POLYNOMIAL_GET_VALUE_FUNCTION_DEFINITION(0)
+
+#define GENERATE_2D_POLYNOMIAL_DIFFERENTIATE_FUNCTION_DEFINITION(FROM_ORDER, TO_ORDER)   \
+    Polynomial2dOrder##TO_ORDER##_t shared_polynomial2d_differentiateOrder##FROM_ORDER(  \
+        Polynomial2dOrder##FROM_ORDER##_t p)                                             \
+    {                                                                                    \
+        Polynomial2dOrder##TO_ORDER##_t derivative = {                                   \
+            .x = shared_polynomial1d_differentiateOrder##FROM_ORDER(p.x),                \
+            .y = shared_polynomial1d_differentiateOrder##FROM_ORDER(p.y),                \
+        };                                                                               \
+        return derivative;                                                               \
+    }
+GENERATE_2D_POLYNOMIAL_DIFFERENTIATE_FUNCTION_DEFINITION(3, 2)
+GENERATE_2D_POLYNOMIAL_DIFFERENTIATE_FUNCTION_DEFINITION(2, 1)
+GENERATE_2D_POLYNOMIAL_DIFFERENTIATE_FUNCTION_DEFINITION(1, 0)
+
+
+#define GENERATE_2D_POLYNOMIAL_GET_ARC_LENGTH_PARAMETRIZATION_FUNCTION_DEFINITION(         \
+    N, N_minus_1)                                                                          \
```

Comment:
Maybe we don't even need to take `N_minus_1` as a macro param, just calculate it in the macro code?

Reply:
Unfortunately not, because we need to use it as a token when we append it to "create" function/type names.

Reply:
Ah. Can we `static_assert` then?

Reply:
Yeah, added.